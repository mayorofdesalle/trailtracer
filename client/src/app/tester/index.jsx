import '@testing-library/jest-dom';

import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';

import './mocks/i18n.mock';
import './mocks/matchMedia.mock';
import './mocks/resizeObserver.mock';

import store from '@app/store';
import GlobalStyle from '@components/style/GlobalStyle';
import Theme from '@components/style/Theme';

const Providers = ({ children }) => {
    return (
        <Provider store={store}>
            <Theme>
                <GlobalStyle />
                <BrowserRouter>
                    {children}
                </BrowserRouter>
            </Theme>
        </Provider>
    );
};

const customRender = (ui, options) =>
    render(ui, { wrapper: Providers, ...options });

// re-export everything
export * from '@testing-library/react';
export * from 'vitest';

// override render method
export { customRender as render };
