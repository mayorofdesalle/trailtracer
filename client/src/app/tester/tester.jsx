import { render } from '@testing-library/react'
import { Provider } from 'react-redux';

import './mocks/matchMedia.mock';
import store from '@app/store';
import Theme from '@components/style/Theme'

const Providers = ({ children }) => {
    return (
        <Provider store={store}>
            <Theme>
                {children}
            </Theme>
        </Provider>
    );
};

const customRender = (ui, options) =>
    render(ui, { wrapper: Providers, ...options });

// re-export everything
export * from '@testing-library/react';
export * from 'vitest';

// override render method
export { customRender as render };