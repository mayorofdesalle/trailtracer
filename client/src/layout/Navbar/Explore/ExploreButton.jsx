import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { set, useForm } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import styled, { css, useTheme } from 'styled-components';

import Icon from '@/components/ui/Icon';
import TextInput from '@/components/ui/input/TextInput';
import NavbarButton from '@/layout/navbar/NavbarButton';
import { FormContext } from '@context';

import ExploreDropdown from './ExploreDropdown';

const ExploreButtonInner = styled(TextInput)`
    width: clamp(3rem, min(6dvw, 6dvh), 12rem);
    height: clamp(1.5rem, min(3dvw, 3dvh), 3rem);
    background-color: ${({ theme }) => theme.colors.primary};
    caret-color: ${({ theme }) => theme.colors.background};
    color: ${({ theme }) => theme.colors.background};
    ${({ $isOpen }) => $isOpen && css`width: 100%;`};

    &:-webkit-autofill,
    &:-webkit-autofill:hover,
    &:-webkit-autofill:focus,
    &:-webkit-autofill:active {
        -webkit-text-fill-color: ${({ theme }) => theme.colors.background};
        caret-color: ${({ theme }) => theme.colors.background};
    }

    &::placeholder {
        color: ${({ theme }) => theme.colors.backgroundTransparent};
    }
`;

/**
 * ExploreButton
 * @description
 * Button that opens a dropdown to explore the site.
 **/
const ExploreButton = () => {
    const theme = useTheme();
    const { t } = useTranslation();

    const { register, handleSubmit, setFocus, watch, setValue } = useForm({
        defaultValues: useMemo(() => { return { query: '' }; }, [])
    });

    const [isSubmitting, setIsSubmitting] = useState(false);

    const onSubmit = (data) => {
        setIsSubmitting(true);
        console.log(data, isSubmitting);
    };

    const dropdown = useRef();
    const [isOpen, setIsOpen] = useState(false);

    const onBlur = useCallback((e) => {
        !(dropdown.current?.contains(e.relatedTarget)
            || Array.from(e.target.parentNode.children).reduce((acc, child) => acc || child.contains(e.relatedTarget), false))
            && setIsOpen(false);
    }, [setFocus]);

    useEffect(() => {
        isOpen && setFocus('search');
    }, [isOpen, setFocus]);

    return (
        <FormContext.Provider value={{ register, setFocus, watch, setValue }}>
            <form onSubmit={handleSubmit(onSubmit)}>
                <ExploreButtonInner as={!isOpen && NavbarButton} $isOpen={isOpen}
                    {...(!isOpen ? { onClick: () => setIsOpen(true), icon: 'menu-search' } : { onBlur: onBlur, type: 'text', name: 'search', clearable: true })}>
                    <Icon name='search-line' color={theme.colors.background} />
                </ExploreButtonInner>
            </form>
            {isOpen && <ExploreDropdown ref={dropdown} close={onBlur} />}
        </FormContext.Provider>
    );
};

export default ExploreButton;