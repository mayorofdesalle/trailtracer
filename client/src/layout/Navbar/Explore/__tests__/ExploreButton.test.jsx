import { describe, expect, it, render, screen, vi } from '@app/tester';

import ExploreButton from '../ExploreButton';

describe('ExploreButton', () => {

    it('should render correctly', () => {
        render(<ExploreButton />);
        const button = screen.getByRole('button');
        expect(button).toBeInTheDocument();

        expect(button.firstChild).toBeInTheDocument();
        expect(button.firstChild).toHaveRole('img');
    });

    it('should match snapshot', () => {
        const { asFragment } = render(<ExploreButton />);
        expect(asFragment()).toMatchSnapshot();
    });

});