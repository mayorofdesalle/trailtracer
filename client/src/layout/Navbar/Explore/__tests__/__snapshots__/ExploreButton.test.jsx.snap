// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExploreButton > should match snapshot 1`] = `
<DocumentFragment>
  <form>
    <button
      class="sc-dmyCSP sc-eTNRI jDIzmn chmNhj sc-ktwOfi eXUhpo"
      type="button"
    >
      <svg
        class="sc-beySPh jIMRxr"
        fill="currentColor"
        role="img"
      >
        <title>
          menu-search-line
        </title>
        <use
          href="http://localhost:3000/src/assets/icons.svg#menu-search-line"
        />
      </svg>
    </button>
  </form>
</DocumentFragment>
`;
